sycl/doc/Assert.md

```diff
@@ -0,0 +1,421 @@
+
+### Interface to program scope variable
+
+Multiple translation units could be compiled/linked into a single device binary
+image. All of them should have `extern` declaration of program scope variable
+available. Definition of the variable is only available within devicelib in the
+same binary image where fallback `__devicelib_assert_fail` resides.
+
+<a name="prog-scope-var-decl">The variable has the following structure and
+declaration:</a>
+
+```c++
+struct AssertHappened {
+  int Flag = 0;
+};
+
+#ifdef __SYCL_DEVICE_ONLY__
+extern SYCL_GLOBAL_VAR AssertHappened AssertHappenedMem;
+#endif
+```
+
+Here, `SYCL_GLOBAL_VAR` is a macro which wraps special attribute to allow for
+mutable program-scope variable.
+
+The reference to extern variable is resolved within online-linking against
+fallback devicelib.
+
+### Online-linking fallback `__devicelib_assert_fail`
+
+Online linking against fallback implementation of `__devicelib_assert_fail` is
+performed only when assertion is enabled and Device-side Runtime doesn't provide
+implementation of `__devicelib_assert_fail`.
+
+In DPCPP headers one can see if assert is enabled with status of `NDEBUG` macro
+with `#ifdef`'s. This allows to enqueue a copy kernel and host task. The copy
+kernel will copy `AssertHappenedMem` to host and host-task will check the `Flag`
+value and `abort()` as needed.
+
+When in DPCPP Runtime Library this knowledge is obtained from device binary
+image descriptor's property sets.
+
+Each device image is supplied with an array of property sets:
+```c++
+struct pi_device_binary_struct {
+  //...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+```
+Each property set is represented by the following struct:
+```c++
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```c++
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+A distinct property set `SYCL/assert used` is added. In this set a property
+with the name of the kernel is added whenever the kernel uses assert. Use of
+assert is detected through call to `__devicelib_assert_fail` function after
+linking device binary image with wrapper device library (the `libsycl-crt`
+library).
+
+The property set and the underlying properties are added by `sycl-post-link`
+tool with help of building callgraph for each and every kernel in device binary
+image.
+
```

Comment:
Besides, it is not always possible to build a callgraph, so, this should be implemented too:
```suggestion
The same is done for all indirectly callable functions (marked with specific attribute) found in the linked device code.
Those are functions whose pointers can be taken and passed around in device code. If a callgraph for any such
function has a call to `__devicelib_assert_fail`, then all kernels in the module are conservatively marked as using asserts.
```

Reply:
@kbobrovs: This is probably not how it will work.  My understanding is that the extension to allow indirect calls in device code is not yet complete or approved.  Last time we talked about this in the language evolution group, we identify ways to narrow down the set of functions that could possibly be called from any one call site.  I expect that all parts of DPC++ that need to form a static call graph will use these techniques to limit the call graph when there are indirect function calls.

Since the indirect function call extension isn't finished yet, it seems better to me that we don't speculate here on how we will handle that case.

Reply:
@gmlueck, in future - maybe, but I'm looking at today's function pointers implementation which works and needs to be covered by this design. Once something changes with function pointers spec, we will update.

Alternatively, the design can mention that the callgraph may be inaccurate in presence of function pointers, and compiler must mark a kernel as "using assert" always, unless it can prove (using future language hints or analysis) that the kernel can never call an indirectly callable function whose (possibly inaccurate) callgraph contains a call to `__devicelib_assert_fail`.

Reply:
Added suggested description.